lib/sessions.js

```diff
@@ -0,0 +1,144 @@
+'use strict';
+
+const retrieveBSON = require('./connection/utils').retrieveBSON,
+  ReadPreference = require('./topologies/read_preference'),
+  BSON = retrieveBSON(),
+  Binary = BSON.Binary,
+  uuidV4 = require('./utils').uuidV4;
+
+/**
+ *
+ */
+class ClientSession {
+  constructor(topology, sessionPool, options) {
+    if (topology == null) {
+      throw new Error('ClientSession requires a topology');
+    }
+
+    if (sessionPool == null || !(sessionPool instanceof ServerSessionPool)) {
+      throw new Error('ClientSession requires a ServerSessionPool');
+    }
+
+    this.topology = topology;
+    this.sessionPool = sessionPool;
+    this.hasEnded = false;
+    this.serverSession = sessionPool.acquire();
+
+    options = options || {};
+    if (typeof options.initialClusterTime !== 'undefined') {
+      this.clusterTime = options.initialClusterTime;
+    }
+  }
+
+  /**
+   *
+   */
+  endSession(callback) {
+    if (this.hasEnded) {
+      return callback(null, null);
+    }
+
+    // TODO:
+    //   When connected to a sharded cluster the endSessions command
+    //   can be sent to any mongos. When connected to a replica set the
+    //   endSessions command MUST be sent to the primary if the primary
+    //   is available, otherwise it MUST be sent to any available secondary.
+    //   Is it enough to use: ReadPreference.primaryPreferred ?
+
+    this.topology.command(
```

Comment:
this is more of a question about the specs - "drivers MUST ignore any errors returned by the endSessions command". ([here](https://github.com/mongodb/specifications/blob/b1ca52d9b7010ccd40e77d9b5848d33e02a4ab37/source/sessions/driver-sessions.rst#endsessions)) why is that the case? also, is this actually ignoring the error if it passes it to a callback? it seems odd that if the endSessions command fails we still mark the session as hasEnded and release it.

Reply:
also, it seems like this command should have the clusterTime with it if it's being gossiped?

Reply:
yeah you're correct about that first point, I shouldn't be passing it up the chain.  The `$clusterTime` is added at the pool level, so it will be included with this command as well :)

Reply:
✅